common/src/main/java/org/conscrypt/OpenSSLEngineImpl.java

```diff
@@ -707,6 +707,12 @@ SSLEngineResult unwrap(final ByteBuffer[] srcs, int srcsOffset, final int srcsLe
                 throw new SSLException("Unable to parse TLS packet header");
             }
 
+            if (packetLength - SSL3_RT_HEADER_LENGTH > capacity) {
+                // No enough space in the destination buffer so signal the caller
+                // that the buffer needs to be increased.
+                return new SSLEngineResult(BUFFER_OVERFLOW, getHandshakeStatus(), 0, 0);
+            }
+
```

Comment:
For this the flow goes:
`inputCiphertext ---write_BIO--> networkBIO buffer --SSL_read--> outputPlaintext`

This is checking `|inputCiphertext|` against `|outputPlaintext|`. However, the actual problem is fitting `inputCiphertext` into `networkBIO buffer`. We should probably use `BIO_ctrl_get_write_guarantee` instead.